src/Microsoft.ML.Core/Utilities/ThreadUtils.cs

```diff
@@ -4,44 +4,154 @@
+            /// <summary>The number of threads currently waiting for work to arrive.</summary>
+            private static int _availableThreads = 0;
+
+            /// <summary>
+            /// Queues an <see cref="Action"/> delegate to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion.  The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task Queue(Action threadStart) => Queue((Delegate)threadStart, null);
+
+            /// <summary>
+            /// Queues an <see cref="Action{Object}"/> delegate and associated state to be executed immediately on another thread,
+            /// and returns a <see cref="Task"/> that represents its eventual completion.  The task will
+            /// always end in the <see cref="TaskStatus.RanToCompletion"/> state; if the delegate throws
+            /// an exception, it'll be allowed to propagate on the thread, crashing the process.
+            /// </summary>
+            public static Task Queue(Action<object> threadStart, object state) => Queue((Delegate)threadStart, state);
+
+            private static Task Queue(Delegate threadStart, object state)
             {
-                IsBackground = false
-            };
+                // Create the TaskCompletionSource used to represent this work.
+                // Call sites only care about completion, not about the distinction between
+                // success and failure and do not expect exceptions to be propagated in this manner,
+                // so only SetResult is used.
+                var tcs = new TaskCompletionSource<bool>(TaskContinuationOptions.RunContinuationsAsynchronously);
+
+                // Queue the work for a thread to pick up.  If no thread is immediately available, create one.
+                if (!Enqueue((threadStart, state, tcs)))
+                {
+                    CreateThread();
+                }
+
+                // Return the task.
+                return tcs.Task;
+
+                void CreateThread()
+                {
+                    // Create a new background thread to run the work.
+                    var t = new Thread(() =>
+                    {
+                        // Repeatedly get the next item and invoke it, setting its TCS when we're done.
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (TryDequeue(out (Delegate, object, TaskCompletionSource<bool>) item))
+                        {
+                            try
+                            {
+                                if (item.Item1 is Action<object> pts)
```

Comment:
>Item1 [](start = 41, length = 5)

I don't object to value tuples per se, but I wonder if the `Item1`, `Item2`, `Item3`, impedes code clarity.

Reply:
It is possible to name them eg `while (TryDequeue(out (Delegate myDelegate, object myObject, TaskCompletionSource<bool> myTcs) item))` and then use those names instead of `Item1` etc.

Reply:
I can name them.